File: src/main/java/ch/obermuhlner/astro/javafx/JavaFXWritableDoubleImage.java
package ch.obermuhlner.astro.javafx;

import ch.obermuhlner.astro.image.color.ColorModel;
import ch.obermuhlner.astro.image.color.ColorUtil;
import ch.obermuhlner.astro.image.DoubleImage;
import javafx.scene.image.WritableImage;

public class JavaFXWritableDoubleImage implements DoubleImage {

  private final WritableImage image;

  public JavaFXWritableDoubleImage(WritableImage image) {
    this.image = image;
  }

  @Override
  public int getWidth() {
    return (int) image.getWidth();
  }

  @Override
  public int getHeight() {
    return (int) image.getHeight();
  }

  @Override
  public double[] getNativePixel(int x, int y, double[] color) {
    if (color == null) {
      color = new double[3];
    }

    int rgb = image.getPixelReader().getArgb(x, y);

    color[ColorModel.RGB.R] = ((rgb >> 16) & 0xff) / 255.0;
    color[ColorModel.RGB.G] = ((rgb >> 8) & 0xff) / 255.0;
    color[ColorModel.RGB.B] = (rgb & 0xff) / 255.0;

    return color;
  }

  @Override
  public void setNativePixel(int x, int y, double[] color) {
    image.getPixelWriter().setArgb(x, y, ColorUtil.toIntARGB(color));
  }

}

File: src/main/java/ch/obermuhlner/astro/javafx/JavaFXWritableDoubleImage.kt
package ch.obermuhlner.astro.javafx

import ch.obermuhlner.astro.image.DoubleImage
import ch.obermuhlner.astro.image.color.ColorModel.RGB
import ch.obermuhlner.astro.image.color.ColorUtil
import javafx.scene.image.WritableImage

class JavaFXWritableDoubleImage constructor(private val image: WritableImage) : DoubleImage {
    override val width: Int
        get() {
            return image.width.toInt()
        }
    override val height: Int
        get() {
            return image.height.toInt()
        }

    override fun getNativePixel(x: Int, y: Int, color: DoubleArray): DoubleArray {
        val rgb: Int = image.pixelReader.getArgb(x, y)
        color[RGB.R] = ((rgb shr 16) and 0xff) / 255.0
        color[RGB.G] = ((rgb shr 8) and 0xff) / 255.0
        color[RGB.B] = (rgb and 0xff) / 255.0
        return color
    }

    override fun setNativePixel(x: Int, y: Int, color: DoubleArray) {
        image.pixelWriter.setArgb(x, y, ColorUtil.toIntARGB(color))
    }
}